File: python/samples/ext.py
try:
    from .test_ctypes import lib, ExtBaseCls, PackedArg, classproperty
except:
    from test_ctypes import lib, ExtBaseCls, PackedArg, classproperty


class ExtTest(ExtBaseCls):
    ext_get = lib.Get("ext_get")
    ext_new = lib.Get("ext_new")
    ext_release = lib.Get("ext_release")

    @classproperty
    def ext_name(cls):
        return "ExtTest"

    @classproperty
    def type_code(cls):
        return 32

    def __init__(self, lib, handle):
        super(ExtTest, self).__init__(lib, handle)

    @property
    def name(self):
        return self.ext_get(self)

    @classmethod
    def _new(cls):
        return cls.ext_new()

    def _release(self):
        self.ext_release(self)
        self.is_py = False

    def __repr__(self):
        return "<%s 0x%x %s>" % (self.ext_name, self.handle, self.name)


PackedArg.RegisterExt(ExtTest)

File: rust/samples/ext.rs
#[macro_use]
extern crate cti;
use cti::*;

pub trait ExtTestType<'lib> : PackedExt<'lib> {
    const _CODE: u32 = 32;
    // TODO: block on default type
    const _NEW: &'static str = "ext_new";
    const _RELEASE: &'static str = "ext_release";
}

impl_packed_ext!(pub struct ExtTest, ExtTestType);
impl_packed_ext!(pub managed struct ExtManagedTest, ExtTestType);

#[derive(Debug)]
pub struct ExtTestProxy<'lib, T> where T: ExtTestBase<'lib> {
    content: T,
    count: u32,
    _phantom: std::marker::PhantomData<&'lib bool>,
}

pub trait ExtTestBase<'lib> : PackedExt<'lib> {
    fn name(&self) -> String;
    fn transform(&mut self) -> ExtTest<'lib>;
}

impl<'lib, T: ExtTestType<'lib>> ExtTestBase<'lib> for T {
    fn name(&self) -> String {
        let f: PackedFunc<'lib> = self.lib().registry_get("PackedFunc", "ext_get");
        packed_call!(f, self)
    }

    fn transform(&mut self) -> ExtTest<'lib> {
        let f: PackedFunc<'lib> = self.lib().registry_get("PackedFunc", "ext_transform");
        packed_call!(f, self as &Self)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    const LIB_NAME: &'static str = "../build/libtest_ctypes.dylib";
    fn _find_lib() -> PathBuf {
        PathBuf::from(LIB_NAME)
    }

    pub fn test_ext() {
        let lib: Lib = Lib::open(_find_lib().as_ref()).unwrap();
        let _ext_test = ExtManagedTest::new(&lib);
        let mut ext_test = ExtTestProxy{ content: _ext_test, count: 2, _phantom: std::marker::PhantomData };
        println!("new_ext: {:?}", ext_test);
        println!("transform: {:?}",
                 (0..ext_test.count).fold(
                     ext_test.content.transform(),
                     |mut t,_| t.transform()));
        let name = ext_test.content.name();
        println!("name: {}", name);
        assert_eq!(name, "run!!!");
    }

    #[test]
    fn it_works() { test_ext() }
}

fn main() {
}
